pkg/controller/sync/version/namespaced.go

```diff
@@ -0,0 +1,162 @@
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+)
+
+type namespacedVersionAdapter struct {
+	sync.RWMutex
+
+	client fedclientset.Interface
+
+	targetKind string
+
+	// Namespace to source propagated versions from
+	namespace string
+
+	versions map[string]*fedv1a1.PropagatedVersion
+}
+
+func newNamespacedVersionAdapter(client fedclientset.Interface, namespace, targetKind string) VersionAdapter {
+	return &namespacedVersionAdapter{
+		client:     client,
+		targetKind: targetKind,
+		namespace:  namespace,
+		versions:   make(map[string]*fedv1a1.PropagatedVersion),
+	}
+}
+
+func (a *namespacedVersionAdapter) TypeName() string {
+	return "PropagatedVersion"
+}
+
+func (a *namespacedVersionAdapter) List(options metav1.ListOptions) (pkgruntime.Object, error) {
+	return a.client.CoreV1alpha1().PropagatedVersions(a.namespace).List(metav1.ListOptions{})
+}
+
+func (a *namespacedVersionAdapter) Iterate(versionList pkgruntime.Object, handler iterationHandler) bool {
+	propagatedVersionList := versionList.(*fedv1a1.PropagatedVersionList)
+	for _, version := range propagatedVersionList.Items {
+		if !handler(&version) {
+			return false
+		}
+	}
+	return true
+}
+
+func (a *namespacedVersionAdapter) GetVersion(qualifiedName util.QualifiedName) pkgruntime.Object {
+	version, _ := a.getVersion(qualifiedName.String())
+	return version
+}
+
+func (a *namespacedVersionAdapter) SetVersion(qualifiedName util.QualifiedName, obj pkgruntime.Object) {
+	version := obj.(*fedv1a1.PropagatedVersion)
+	a.setVersion(qualifiedName.String(), version)
+}
+
+func (a *namespacedVersionAdapter) DeleteVersion(qualifiedName util.QualifiedName) error {
+	err := a.client.CoreV1alpha1().PropagatedVersions(qualifiedName.Namespace).Delete(qualifiedName.Name, nil)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	a.Lock()
+	delete(a.versions, qualifiedName.String())
+	a.Unlock()
+	return nil
+}
+
+func (a *namespacedVersionAdapter) GetStatus(qualifiedName util.QualifiedName) (*fedv1a1.PropagatedVersionStatus, bool) {
+	a.RLock()
+	version, ok := a.versions[qualifiedName.String()]
+	a.RUnlock()
+	if !ok {
+		return nil, ok
+	}
+	status := version.Status
+	return &status, ok
+}
+
+func (a *namespacedVersionAdapter) SetStatus(qualifiedName util.QualifiedName, status *fedv1a1.PropagatedVersionStatus) {
+	key := qualifiedName.String()
+	version, ok := a.getVersion(key)
+	if ok {
+		version.Status = *status
+	} else {
+		version = &fedv1a1.PropagatedVersion{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: qualifiedName.Namespace,
+				Name:      qualifiedName.Name,
+			},
+			Status: *status,
+		}
+		a.setVersion(key, version)
```

Comment:
Good catch.  I think I've been glossing over how modifications to map entries should be synchronized.  I should really be treating map values as immutable and always calling the mutator methods instead of blithely modifying the version objects directly.